keps/sig-autoscaling/2021-scale-from-zero/README.md

```diff
@@ -0,0 +1,939 @@
+with and without the feature, are necessary. At the very least, think about
+conversion tests if API types are being modified.
+-->
+
+There currently unit tests for the alpha cases and tests planned to be added for the new functionality.
+
+### Rollout, Upgrade and Rollback Planning
+
+<!--
+This section must be completed when targeting beta to a release.
+-->
+
+As this is a new field every usage is opt-in. In case the kubernetes version is downgraded, currently scaled to 0 workloads might not to be manually scaled to 1 as the controller would treat them as
+paused otherwise.
+
+###### How can a rollout or rollback fail? Can it impact already running workloads?
+
+<!--
+Try to be as paranoid as possible - e.g., what if some components will restart
+mid-rollout?
+
+Be sure to consider highly-available clusters, where, for example,
+feature flags will be enabled on some API servers and not others during the
+rollout. Similarly, consider large clusters and how enablement/disablement
+will rollout across nodes.
+-->
+
+There are no expected side-effects when the rollout fails as new `enableScaleToZero` flag should only be enabled once the version upgraded completed and should be disabled before attempting a rollback.
+
+In case this is missed, HPA for deployments with zero replicas will be disabled until replicas have been raised explicitly to at least `1`.
+
+###### What specific metrics should inform a rollback?
+
+<!--
+What signals should users be paying attention to when the feature is young
+that might indicate a serious problem?
+-->
+
+If workloads aren't scaled up from 0 despite the scaling condition being meet, an operator should rollback this feature and manually scale an affected workload back to `1`.
+
+###### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?
+
+<!--
+Describe manual testing that was done and the outcomes.
+Longer term, we may want to require automated upgrade/rollback tests, but we
+are missing a bunch of machinery and tooling and can't do that now.
+-->
+
+No yet as no implementation is available.
+
+###### Is the rollout accompanied by any deprecations and/or removals of features, APIs, fields of API types, flags, etc.?
+
+<!--
+Even if applying deprecation policies, they may still surprise some users.
+-->
+
+Once this is in beta, the alpha flag can be removed.
+
+### Monitoring Requirements
+
+<!--
+This section must be completed when targeting beta to a release.
+
+For GA, this section is required: approvers should be able to confirm the
+previous answers based on experience in the field.
+-->
+
+###### How can an operator determine if the feature is in use by workloads?
+
+<!--
+Ideally, this should be a metric. Operations against the Kubernetes API (e.g.,
+checking if there are objects with field X set) may be a last resort. Avoid
+logs or events for this purpose.
+-->
+
+The feature is used if workloads are scaled to zero by the autoscaling controller.
```

Comment:
Can we provide a one liner for operators to examine the HPAs in their clusters for opt-in use of the `enableScaleToZero` field?

Reply:
If you have 50,000 clusters, a one liner is not really helpful. Metrics are best. Can kube state metrics help?

If it's not reasonable/practical to have a metric, then other means can be OK.

Reply:
Yes, the property could be added to https://github.com/kubernetes/kube-state-metrics/blob/main/internal/store/horizontalpodautoscaler.go as an additional `kube_horizontalpodautoscaler_` metric.